src/wallet/silentpayments.cpp

```diff
@@ -0,0 +1,294 @@
+                        CPubKey labeled_spend_key{P_k + m_G};
+                        if (output_pubkey.ConvertToCompressedPubKey(/*even=*/true) == labeled_spend_key) {
+                            t_k = combine_tweaks(t_k, m);
+                            found = true;
+                            break;
+                        }
+                        // If we don't find a label with the even output, also check the odd (negated) output
+                        // Note: changing the parity byte from even to odd is not as expensive as negating an EC point
+                        if (output_pubkey.ConvertToCompressedPubKey(/*even=*/false) == labeled_spend_key) {
+                            t_k = combine_tweaks(t_k, m);
+                            found = true;
+                            break;
+                        }
+                    }
+                } else {
+                    // If we have 3 or more labels, checking via subtraction is cheaper.
+                    // This ends up being at most N * 2 ECC operations (two subtractions) and a lookup in the labels map,
+                    // where N is the number of taproot outputs in the transaction
+                    std::map<CPubKey, uint256>::const_iterator it = labels.end();
+                    // For an XOnlyPubKey, converting it to an odd compressed key is equivalent to negation.
+                    // We use P_k_negated for subtraction: output + ( - P_k ),
+                    // so first we need to check the parity byte of the compute P_K public key
+                    bool parity = (P_k.begin()[0] == 0x02) ? true : false;
+                    CPubKey P_k_negated = P_k_xonly.ConvertToCompressedPubKey(/*even=*/!parity);
+                    it = labels.find(output_pubkey.ConvertToCompressedPubKey(/*even=*/true) + P_k_negated);
+                    if (it == labels.end()) {
+                        // If not found on the first try, negate the output and check again.
+                        it = labels.find(output_pubkey.ConvertToCompressedPubKey(/*even=*/false) + P_k_negated);
+                    }
+                    if (it != labels.end()) {
+                        t_k = combine_tweaks(t_k, it->second);
+                        found = true;
+                    }
+                }
+            }
+            if (found) {
+                // Since we found an output, we need to increment k and check the vector again
+                tweaks.emplace_back(t_k);
+                keep_going = true;
+                k++;
+                // Return true so that this output pubkey is removed the from vector and not checked again
+                return true;
+            }
+            return false;
+        }), output_pub_keys.end());
+    } while (!output_pub_keys.empty() && keep_going);
+    return tweaks;
+}
+
+std::optional<std::pair<uint256, CPubKey>> GetSilentPaymentsTweakDataFromTxInputs(const std::vector<CTxIn>& vin, const std::map<COutPoint, Coin>& coins)
```

Comment:
According to the BIP we should skip the whole tx if there's unknown witness input. As currently, implemented it will silently skip just that input and attempt to proceed. I added a test vector for unknown witness version